shell.nix

```diff
@@ -0,0 +1,112 @@
+# Licensed under the Apache License, Version 2.0 or the MIT License.
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+# Copyright Tock Contributors 2022.
+
+# Shell expression for the Nix package manager
+#
+# This nix expression creates an environment with necessary packages installed:
+#
+#  * `tockloader`
+#  * rust
+#
+# To use:
+#
+#  $ nix-shell
+#
+
+{ pkgs ? import <nixpkgs> { }, withUnfreePkgs ? false }:
+
+with builtins;
+let
+  inherit (pkgs) stdenv lib;
+
+  # Tockloader v1.11.0
+  tockloader = import (pkgs.fetchFromGitHub {
+    owner = "tock";
+    repo = "tockloader";
+    rev = "v1.11.0";
+    sha256 = "sha256-bPEfpfOZOjOiazqRgn1cnqe4ohLPvocuENKoZx/Qw80=";
+  }) { inherit pkgs withUnfreePkgs; };
+
+  # Rust toolchain overlay
+  rust_overlay = import "${pkgs.fetchFromGitHub {
+    owner = "nix-community";
+    repo = "fenix";
+    # Revision hash must correspond with the same date as the nightly version.
+    rev = "ffa0a8815be591767f82d42c63d88bfa4026a967"; 
+    sha256 = "sha256-JOrXleSdEKuymCyxg7P4GTTATDhBdfeyWcd1qQQlIYw=";
+  }}/overlay.nix";
+
+  nixpkgs = import <nixpkgs> { overlays = [ rust_overlay ]; };
+  stable = (lib.importTOML ./rust-toolchain.toml).toolchain;
+  nightly = (lib.importTOML ./nightly/rust-toolchain.toml).toolchain;
+
+  # Use nightly development toolchain because Miri is not supported
+  # by the MSRV (Minimum Supported Rust Version) toolchain.
+  nightlyToolchain = nixpkgs.fenix.fromToolchainName {
+    name = nightly.channel;
+    sha256 = "sha256-R/ONZzJaWQr0pl5RoXFIbnxIE3m6oJWy/rr2W0wXQHQ=";
+  };
```

Comment:
I appreciate your efforts in making this work by combining the two toolchains, and while I think that this is a decent approach generally, to me this really points to a deficiency in how we pick nightly toolchain versions for `libtock-rs`. For the Tock kernel, we have a script that searches for a Nightly that has all of the required components (including Miri). I think we should rather choose a proper nightly this way, and avoid mix- and matching Rust components. I'm not even sure whether Miri is portable across rustc versions (e.g., in the features that it recognizes).

Reply:
I removed the dual toolchain installation from before. Only the nightly derivation is generated, with the references on [lines 41-42](https://github.com/tock/libtock-rs/pull/519/files#diff-e53dfbfffe62ae3c0b411b3938ccffa9fb6a2ecc565f55785ef8daa756631a6bR41-R42) being to the pure text of the file. This Nix logic has the (virtually) same effect as applying the diff in my original post before doing the toolchain derivation generation.
The problem you foresaw with some components being unavailable was not an issue. `make test` works, but I agree that incompatible nightlies are a risk in the future.

Reply:
I'm sorry, I'm having trouble matching what you write with the current Nix expression of this PR. Lines [lines 51-57](https://github.com/tock/libtock-rs/pull/519/files#diff-e53dfbfffe62ae3c0b411b3938ccffa9fb6a2ecc565f55785ef8daa756631a6bR51-R57) still provide you a toolchain that is efffectively mix-and-match between stable and binary, and what you get is dependent on which components these two (stable + nightly) toolchains happen to provide, right?

And  [lines 58-65](https://github.com/tock/libtock-rs/pull/519/files#diff-e53dfbfffe62ae3c0b411b3938ccffa9fb6a2ecc565f55785ef8daa756631a6bR58-R65) then also throw in some components (namely, `rust-std`) taken from whatever the fenix overlay defines as the latest toolchain, which only happens to coincide with the pinned nightly toolchain by the correctness of the condition in [lines 35](https://github.com/tock/libtock-rs/pull/519/files#diff-e53dfbfffe62ae3c0b411b3938ccffa9fb6a2ecc565f55785ef8daa756631a6bR35)? That seems very brittle.